---
title: "Step2b_data_processing"
author: "T. Divoll"
date: "2023-09-12"
output: 
  html_document:
  df_print: paged
  toc: true
---

```{r setup, include=FALSE}
# set global chunk parameters here
knitr::opts_chunk$set(echo = TRUE)
```

#### Merge Illumina forward and reverse reads

```{bash}
#Recover full sequence reads from forward and reverse partial reads using illuminapiredend command in obitools

for sample in $(cat sampleNames)
do
    echo "On sample: $sample"
    
    illuminapairedend --score-min=$matching_score \
    -r ${sample}_S[0-9]+_L001_R2_001_trimmed.fastq.gz \
    ${sample}_S[0-9]+_L001_R1_001_trimmed.fastq.gz > ${sample}_merged.fastq
    
    echo obicount ${sample}_merged.fastq
done

```

#### Filter out unaligned reads
```{bash}
for sample in $(cat sampleNames)
do

    echo "On sample: $sample"
    
    obigrep -p 'mode!="joined"' ${sample}_merged.fastq > ${sample}_ali.fastq
done
```

#### Add sample names to each file
```{bash}
for sample in $(cat sampleNames)
do

    echo "On sample: $sample"
    
  obiannotate -S sample:${sample} ${sample}_ali.fastq >  ${sample}_id.fastq

done

mv *_id.fastq $merged_dir
```

#### Combine remaining sequences into one FASTA and dereplicate
```{bash}
cd $merged_dir

obiuniq -m sample *_id.fastq > all.uniq.fasta
echo obicount all.uniq.fasta
```

