---
title: "Step 2a - data preparation"
author: "T. Divoll"
date: "2023-09-12"
output: 
  html_document:
  df_print: paged
  toc: true
---

```{r setup, include=FALSE}
# set global chunk parameters here
knitr::opts_chunk$set(echo = TRUE)
```

This notebook can be used to run the OBITools 1.2.12 pipeline and serve as a tutorial. Run through each code chunk and inspect the outputs. First, update the parameters in the first few code chunks.

```{r include=FALSE}
if (!require("pacman")) install.packages("pacman")
if (!require("BiocManager", quietly = TRUE))
pacman::p_load(here, tidyverse, readxl, stringr, Biostrings)
here::i_am(file.path(here(today, "src/Step2a_data_prep.Rmd")))
```

#### Update parameters and thresholds

```{r}
# OBITools parameters
min_read_count <- 1000
min_read_length <- 8 # -l in obigrep
max_read_length <- 300 # -L in obigrep
clean_threshold <- 0.05 # -r in obiclean
matching_score <- 40 # --score-min in illuminapairedend

# Cutadapt parameters
cutadapt_mismatch <- 2 # -e in cutadapt
F_primer <- DNAString("GGGCAATCCTGAGCCAA")
R_primer <- DNAString("CCATTGAGTCTCTGCACCTATC")

## make environment variables that can be passed to bash code chunks
Sys.setenv(cutadapt_mismatch = cutadapt_mismatch)
Sys.setenv(min_read_count = min_read_count)
Sys.setenv(min_read_length = min_read_length)
Sys.setenv(max_read_length = max_read_length)
Sys.setenv(clean_threshold = clean_threshold)
Sys.setenv(matching_score = matching_score)
Sys.setenv(F_primer = F_primer)
Sys.setenv(R_primer = R_primer)
```

#### Generate some additional parameters from inputs

```{r}
F_primer_reverse <- reverseComplement(F_primer)
R_primer_reverse <- reverseComplement(R_primer)
F_primer_reverse 
R_primer_reverse

## make environment variables that can be passed to bash code chunks

Sys.setenv(F_primer_reverse = F_primer_reverse)
Sys.setenv(R_primer_reverse = R_primer_reverse)
```

The following dates are pulled from the sample sheet and can be used to filter out dates in case of contamination. The unique extraction, PCR, and sequencing dates are cast to lists that we can then index into later.

```{r}
extraction_date <- as.list(unique(sample_sheet$`Ext date`))
pcr_date <- as.list(unique(sample_sheet$`PCR date`))
sequencing_date <- as.list(unique(sample_sheet$`Seq date`))
```

#### Rename files with project-specific sample names

We can make a list of the existing fasta.gz files, and then use column pairings with lab and field sample names to update the file names.

```{r}
file_names_original <- list.files(samples_dir, pattern=".gz", full.names = T) 
lab_name <- sample_sheet$SampleName
field_name <- sample_sheet$SampleID
df <- data.frame(field_name, lab_name)
file_names <- stringr::str_replace_all(file_names_original, setNames(df$field_name, df$lab_name))
file.rename(file_names_original, file_names)
```

```{r}
control_samples <- sample_sheet %>% 
  filter(Control == "Y")
control_samples <- unique(control_samples$SampleID)
```


```{r}
capture.output(control_samples, file = here(today, 'results/control_list.txt'))
capture.output(file_names, file = here(today, 'results/file_names.txt'))
```


```{r}
sampleNames <- as.list(field_name)
Sys.setenv(sampleNames = "sampleNames")
```

#### Trim primers with Cutadapt

This section will strip off the primers and send the "stdout" to a file so we can keep track of what was removed for each sample.

**NOTE** At this point, the code chunks switch over from R to Bash

```{bash}
# installing cutadapt - create environment to install into
conda create -n cutadaptenv
conda activate cutadaptenv
conda install -c bioconda cutadapt
```

```{bash}
cd output_dir

for sample in $(cat sampleNames)
do

    echo "On sample: $sample"
  
    cutadapt -a ^$F_primer...$R_primer_reverse -A ^$R_primer...$R_primer_reverse --no-indels -e $cutadapt_mismatch --discard-untrimmed -o ${sample}_S[0-9]+_L001_R1_001_trimmed.fastq.gz -p ${sample}_S[0-9]+_L001_R2_001_trimmed.fastq.gz ${sample}_S[0-9]+_L001_R1_001.fastq.gz ${sample}_S[0-9]+_L001_R2_001.fastq.gz >> cutadapt_primer_trimming_stats.txt 2>&1
done
```

Now look at what fraction of reads were retained in each sample (column 2) and what fraction of bps were retained in each sample (column 3). Expect \~95%+ of reads to be retained and \~50-70% of bps.

```{bash}
paste sampleNames <(grep "passing" cutadapt_primer_trimming_stats.txt | cut -f3 -d "(" | tr -d ")") <(grep "filtered" cutadapt_primer_trimming_stats.txt | cut -f3 -d "(" | tr -d ")")
```

Deactivate the conda environment that uses Python 3 for Cutadapt

```{bash}
conda deactivate
```

Now navigate to the notebook `Step2b_data_processing.Rmd` in the Files window pane. All of the variables and parameters should still be available in the R environment (see the `Environment` window pane.)
