---
title: "Step2b_data_processing"
author: "T. Divoll"
date: "2023-09-12"
output: 
  html_document:
  df_print: paged
  toc: true
---
```{r include=FALSE}
## Only needed for knitr to render this notebook
if (!require("pacman")) install.packages("pacman")
pacman::p_load(knitr, here, tidyr, dplyr, readxl, stringr, spgs, filesstrings)
```

#### Set up Conda env for OBITools

```{bash, install obitools, eval = FALSE, echo = FALSE}
# this only needs to be run once => skip if you have already created this in a prior analysis
source ~/.bash_profile
conda create -n obitools-env -y -c conda-forge -c bioconda obitools ecoprimers ecopcr python=2.7
```

#### Merge Illumina forward and reverse reads

```{bash}
#Recover full sequence reads from forward and reverse partial reads using illuminapiredend command in obitools
source ~/.bash_profile
conda activate /users/${user}/.conda/envs/obitools-env
cd $output_path

for sample in $(cat $samples_path/sampleNames)
do
    echo "On sample: $sample"
    
    illuminapairedend --score-min=$matching_score -r ${sample}_L001_R2_001_trimmed.fastq.gz ${sample}_L001_R1_001_trimmed.fastq.gz > ${sample}_merged.fastq
    
    obicount ${sample}_merged.fastq
done

```

#### Filter out unaligned reads
```{bash}
source ~/.bash_profile
conda activate /users/${user}/.conda/envs/obitools-env
cd $output_path

for sample in $(cat $samples_path/sampleNames)

do
    obigrep -p 'mode!="joined"' ${sample}_merged.fastq > ${sample}_ali.fastq
done | column -t
```

#### Add sample names to each file
```{bash}
source ~/.bash_profile
conda activate /users/${user}/.conda/envs/obitools-env
cd $output_path

for sample in $(cat $samples_path/sampleNames)

do
    obiannotate -S sample:${sample} ${sample}_ali.fastq >  ${sample}_id.fastq

done | column -t

mv *_id.fastq $merged_path
```

#### Inspect read counts in fecal samples vs controls
```{bash}
source ~/.bash_profile
conda activate /users/${user}/.conda/envs/obitools-env
cd $merged_path

obistat -c sample *_id.fastq --without-progress-bar | column -t
```

#### Combine remaining sequences into one FASTA and dereplicate
```{bash quiet=TRUE}
source ~/.bash_profile
conda activate /users/${user}/.conda/envs/obitools-env
cd $merged_path

obiuniq -m sample *_id.fastq --without-progress-bar > all.uniq.fasta

obicount all.uniq.fasta --without-progress-bar
```

#### Filter results to just `count` and `merged_sample` attributes in FASTA headers
```{bash}
source ~/.bash_profile
conda activate /users/${user}/.conda/envs/obitools-env
cd $merged_path

obiannotate -k count -k merged_sample all.uniq.fasta > $$ ; mv $$ all.uniq.fasta
```

#### Inspect the `count` attribute
```{bash}
source ~/.bash_profile
conda activate /users/${user}/.conda/envs/obitools-env
cd $merged_path

# get the counting statistics on the ‘count’ attribute
obistat -c count all.uniq.fasta --without-progress-bar | sort -nk1 | head -20 | column -t
```

#### Create a tabulated file for counts per sequence read
```{bash}
source ~/.bash_profile
conda activate /users/${user}/.conda/envs/obitools-env
cd $merged_path

obitab -o all.uniq.fasta > all.uniq.tab
```
#### Read the tab file as a table in R
After running the cell below, open the table from the Environment tab and inspect the results!
```{r}
all_uniq_tab <- read.delim(here(merged_path, "all.uniq.tab"), header=TRUE)
```

Use this space to take notes on the controls that have been looked into. For example, something like this: 
```
BLA20220920.A
6671/17628 (38%) – Glehnia littoralis = American silvertop
3442/5438 (63%) – Musa balbisiana = plantain
```
#### Move all the control samples over to an archive directory
Now that the controls have been looked into, we can move the sample files over to a folder called `control_samples`
```{r}
merged_files <- list.files(here(params$datetime, "results/merged"), pattern="*id.fastq", full.names = T)
```

```{r}
control_files <- merged_files[grepl(paste(control_samples, collapse= "|"), merged_files)]
```

Add all of the `*id.fastq` files you want to move to the list named `bad_samples_to_move`. This will move them to a `bad_controls_samples` folder. These can be from suspected contamination or from low read counts (i.e., < 1000 reads).
```{r}
bad_samples_to_move <- c("", "")
```

```{r}
move_files(control_files, bad_controls_path, overwrite = FALSE)
move_files(bad_samples_to_move, bad_controls_path, overwrite = FALSE)
```

Let's also move this first data set to archive and we will re-run merging and de-replication in the next notebook, without controls and bad samples.
```{r}
move_files(c(here(merged_path, "all.uniq.fasta"), here(merged_path, "all.uniq.tab")), bad_controls_path, overwrite = FALSE)
```
Now navigate to the notebook `Step2c_data_cleaning.Rmd` in the Files window pane. All of the variables and parameters should still be available in the R environment (see the `Environment` window pane.)